eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/CDDL.hs

```diff
@@ -0,0 +1,1008 @@
+      , a operational_cert
+      , a protocol_version
+      ]
+
+operational_cert :: Rule
+operational_cert =
+  "operational_cert"
+    =:= arr
+      [ "hot_vkey" ==> kes_vkey
+      , "sequence_number" ==> (VUInt `sized` (8 :: Word64))
+      , "kes_period" ==> VUInt
+      , "sigma" ==> signature
+      ]
+
+protocol_version :: Rule
+protocol_version = "protocol_version" =:= arr [a major_protocol_version, a VUInt]
+
+-- TODO Replace with the following once
+-- https://github.com/input-output-hk/cuddle/issues/29 is addressed in cuddle.
+--
+-- next_major_protocol_version :: Rule
+-- next_major_protocol_version = "next_major_protocol_version" =:= (10 :: Integer)
+next_major_protocol_version :: Integer
+next_major_protocol_version = 10
+
+major_protocol_version :: Rule
+major_protocol_version = "major_protocol_version" =:= (1 :: Integer) ... next_major_protocol_version
+
+transaction_body :: Rule
+transaction_body =
+  "transaction_body"
+    =:= mp
+      [ idx 0 ==> set transaction_input
+      , idx 1 ==> arr [0 <+ a transaction_output]
+      , idx 2 ==> coin
+      , opt (idx 3 ==> slot_no)
+      , opt (idx 4 ==> certificates)
+      , opt (idx 5 ==> withdrawals)
+      , opt (idx 7 ==> auxiliary_data_hash)
+      , opt (idx 8 ==> slot_no) -- Validity interval start
+      , opt (idx 9 ==> mint)
+      , opt (idx 11 ==> script_data_hash)
+      , opt (idx 13 ==> nonempty_set transaction_input)
+      , opt (idx 14 ==> required_signers)
+      , opt (idx 15 ==> network_id)
+      , opt (idx 16 ==> transaction_output)
+      , opt (idx 17 ==> coin)
+      , opt (idx 18 ==> nonempty_set transaction_input)
+      , opt (idx 19 ==> voting_procedures)
+      , opt (idx 20 ==> proposal_procedures)
+      , opt (idx 21 ==> coin)
+      , opt (idx 22 ==> positive_coin)
+      ]
+
+voting_procedures :: Rule
+voting_procedures =
+  "voting_procedures"
+    =:= mp [1 <+ asKey voter ==> mp [1 <+ asKey gov_action_id ==> voting_procedure]]
+
+voting_procedure :: Rule
+voting_procedure = "voting_procedure" =:= arr [a vote, a anchor / VNil]
+
+proposal_procedure :: Rule
+proposal_procedure =
+  "proposal_procedure"
+    =:= arr
+      [ "deposit" ==> coin
+      , a reward_account
+      , a gov_action
+      , a anchor
+      ]
+
+proposal_procedures :: Rule
+proposal_procedures = "proposal_procedures" =:= nonempty_set proposal_procedure
+
+certificates :: Rule
+certificates = "certificates" =:= nonempty_set certificate
+
+gov_action :: Rule
+gov_action =
+  "gov_action"
+    =:= arr
+      [ a parameter_change_action
+          / hard_fork_initiation_action
+          / treasury_withdrawals_action
+          / no_confidence
+          / update_committee
+          / new_constitution
+          / info_action
+      ]
+
+policy_hash :: Rule
+policy_hash = "policy_hash" =:= scripthash
+
+parameter_change_action :: Named Group
+parameter_change_action =
+  "parameter_change_action"
+    =:~ grp
+      [ 0
+      , gov_action_id // VNil
```

Comment:
So, in CDDL, `/` is a choice between types (concretely, between `Type1` values, to make a `Type0`). `//` allows choice between groups. We can illustrate the difference:

```
foo = [ 0 / 1, uint // 2 /3, tstr ]
```
This construction would match either of the following:
```
[0, 3]
[2, "Hello World"]
```

 In other words, the `//` binds less strongly than `,` in CDDL.

In Haskell, of course, we cannot have syntax inside an array which binds stronger than the `,`. so we have to do things a little differently. The way this is handled at the moment is that `//` has special treatment for arrays/groups, where it will, instead of creating a type-level choice, merge the two arrays/groups/maps into a single one containing a group choice.

So, two questions remain: why is this operator `//` instead of `/` (whose semantics it more closely matches), and what is the `/` operator in Huddle?

The former is largely accidental: the group choice operator got expanded to other cases.

What about the latter? Well, that was mostly intended to handle the case of array entries. But looking at it again, I think it's a mistake. The case it was used for can be replaced by updating fixity declarations.

So I think I will update this with the following intention: all choices within huddle will be written with `/`, and I will incorporate most of this comment into the documentation to clarify it.